### ISSUE TITLE
Modelica.Media.Air.MoistAir leads to illegal Modelica lookup!

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#482]
**Reported by adrpo on 19 Jan 2011 06:28 UTC**
We start from definition of MoistAir:
```
package MoistAir "Air: Moist air model (240 ... 400 K)"
  extends Interfaces.PartialCondensingGases(
    mediumName="Moist air",
    substanceNames={"water", "air"},
    final reducedX=true,
    final singleState=false,
    reference_X={0.01,0.99},
    fluidConstants = {IdealGases.Common.FluidData.H2O,IdealGases.Common.FluidData.N2});
   ....
  end MoistAir;
```

So, in MoistAir we have constant:
```
IdealGases.Common.FluidData.H2O
```

The constant is defined in Modelica.Media.IdealGases.Common as:
```
constant SingleGasNasa.FluidConstants H2O(
  chemicalFormula =        "H2O",
  iupacName =              "unknown",
  structureFormula =       "unknown",
  casRegistryNumber =      "7732-18-5",
  meltingPoint =           273.15,
  normalBoilingPoint =     373.15,
  criticalTemperature =    647.14,
  criticalPressure =       220.64e5,
  criticalMolarVolume =     55.95e-6,
  acentricFactor =           0.344,
  dipoleMoment =             1.8,
  molarMass =              SingleGasesData.H2O.MM,
  hasDipoleMoment =       true,
  hasIdealGasHeatCapacity=true,
  hasCriticalData =       true,
  hasAcentricFactor =     true);
```

So far so good, but now the problems begin: SingleGasNasa is a partial package!
In Modelica Specification 3.2, section 5.3.2 Composite Name Lookup is said:
"The class we look inside may not be partial in a simulation model."
So, accessing SingleGasNasa.FluidConstants is illegal because SingleGasNasa is partial!

Besides this (to be even weirder) the record SingleGasNasa.FluidConstants is a record declared as:
```
 redeclare record extends FluidConstants
   ...
 end FluidConstants;
```
so you actually need to instantiate SingleGasNasa, apply the extends in it,
but you can't do that because it uses data.name which has no binding!
```
partial package SingleGasNasa
  "Medium model of an ideal gas based on NASA source"
  extends Interfaces.PartialPureSubstance(
     ThermoStates = Choices.IndependentVariables.pT,
     mediumName=data.name,
     substanceNames={data.name},
     singleState=false,
     Temperature(min=200, max=6000, start=500, nominal=500),
     SpecificEnthalpy(start=if referenceChoice==ReferenceEnthalpy.ZeroAt0K then data.H0 else
        if referenceChoice==ReferenceEnthalpy.UserDefined then h_offset else 0, nominal=1.0e5),
     Density(start=10, nominal=10),
     AbsolutePressure(start=10e5, nominal=10e5));
 
   constant IdealGases.Common.DataRecord data "Data record of ideal gas substance";
end SingleGasNasa;
```

So, how does Modelica.Media works in Dymola if is not legal Modelica?

Of course, I might be wrong and if I am please enlighten me!

Cheers,
Adrian Pop/

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/482

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 21 Jan 2011 21:11 UTC**
I'd like to know if you think the problem is with the construct as used, or whether the problem is with the clarity of the specification. To me it is not strange do define the type FluidConstants in a partial package. The trouble may be with the "everything is a class" notion that treats packages and models (mostly) the same in the specification. The package SingleGasNasa is in my opinion not "in a simulation model" ("The class we look inside may not be partial in a simulation model."). It does not need to be instantiated, and I don't know a reason why a package should ever need to be instantiated. In this particular case it may be possible to move the declaration of FluidConstants to another package, but I think hte problem is rather with the specification than with the usage.

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 21 Jan 2011 22:41 UTC**
The problem is that the record SingleGasNasa.FluidConstants is defined with
`redeclare record extends` (which can both add components and modify an existing class) which means that if you want to construct is full type you 
need to handle the extends in SingleGasNasa (and all the way up to PartialMedium) which means that you need to flatten (instantiate SingleGasNasa) which is partial (and also the extends uses data.name which has no defined binding).

The problem with the specification (or our interpretation of it) is that 
it creates the impression that a partial class/package/\* cannot be used
(via dot notation) if is not modified (via extends), but in this case 
you have such a usage. I don't know if there are other usages of a 
partial package/class/\* via the dot notation inside MSL (or Modelica
Specification), most of them are through extends.

We can probably solve this in OpenModelica by changing a bit the way we handle flattening of packages (SingleGasNasa) to allow it even if modifications in extends (data.name) have no binding and then only verify after the flattening
(in the DAE) if we have all the bindings we need. Is not very clean because is
far better to detect an error or a problem early (not after the flattening).

I think this issue (usage of partial package via dot notation) 
could be solved if inside Modelica.Media.IdealGases.Common.FluidData
instead of package SingleGasNasa[.FluidConstants] we use another package 
that extends SingleGasNasa and sets all the needed things as is done in:
Modelica.Media.IdealGases.SingleGases.H2O for example. Note that we cannot 
use this already existing package as we would get a circular dependency on constant Common.FluidData.H2O.

I also think that the Modelica specification should make it more clear
that you don't need to have the full information (constant bindings, etc) 
for some constructs (for example structural parameters need to have the
full information) in order to flatten/instantiate a class. From an 
implementation point of view this means that you need \*very\* fine grain 
dependency analysis: keep ONLY what is used and disregard the rest without 
checking, which can be bad as it doesn't detect some errors. 

Consider the text "Components declared as constant shall have an associated declaration equation with a constant expression, if the constant is used in 
the (simulation) model." What does "is used" means here:
- appears a binding or inside a binding expression? (if you  have it as a binding for another constant that is not used does it means that both are unused?)
- appears inside an equation or algorithm?
- how about a type? Real(min=constantWithNoBinding)

Cheers,
Adrian Pop/

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 24 Jan 2011 14:48 UTC**
Discussion at the 69th design meeting:

The current design of the Media library is not nice, because the large “FluidConstant” record is defined at 3 places with identical definitions:

 Media.Interfaces.PartialMixtureMedium,
 Media.Interfaces.PartialTwoPhaseMedium,
 Modelica.Media.IdealGases.Common.SingleGasNasa.FluidConstants

These definitions use the type definitions of the respective medium, such as:

 Temperature
 AbsolutePressure

Solution proposal to change the Media library:

1. Introduce the large “FluidConstant” record in Modelica.Media.IdealGases.Common as a new record (that does not depend on anything else).

2. In package Modelica.Media.IdealGases.Common.FluidData replace all decalartions of the form:
 constant SingleGasNasa.FluidConstants N2
by
 constant Modelica.Media.IdealGases.Common.FluidConstants N2
As a result, the FluidConstant data is in a non-partial package and is independent from other packages

3. At all places where FluidConstant is defined, change the definition by extending from
 Modelica.Media.IdealGases.Common.FluidConstants
This change is backwards compatible with the only exception, that “SI.Temperature” etc. is used, and not the medium specific type definition. However, this should not be an issue for a constant.

In order to test this with OpenModelica, MSL3.1 needs to be used. The proposal is therefore to make a branch from the latest maintenance release of MSL3.1, change this version according to the proposal above and let Adrian Pop test it with OpenModelica. 

Hubertus plans to perform this during the design meeting.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 25 Jan 2011 14:02 UTC**
Adrian: I changed the definition of FluidConstants in the branch https://svn.modelica.org/projects/Modelica/branches/maintenance/3.1-MediaFix. It works fine in Dymola, I hope it makes it easier for you. Let me know if you think this is better, then I'll carry the change over to trunk as well.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 28 May 2012 22:29 UTC**
Closing now this one: still needs to be merged to trunk with a number of related tickets.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 29 May 2012 06:02 UTC**
There is a danger of forgetting about this and also for the sake of documentation we should leave it open until we know for which Milestone it was merged.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 10 Jun 2012 23:16 UTC**
fixed in a251c81439d91417bda21fba4ef0f945a3854365
